crates/evm/src/borsh_compat.rs

```diff
@@ -0,0 +1,531 @@
+    }
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+enum CompatTransaction {
+    Legacy(CompatTxLegacy),
+    Eip2930(CompatTxEip2930),
+    Eip1559(CompatTxEip1559),
+    Eip4844(CompatTxEip4844),
+    Eip7702(CompatTxEip7702),
+}
+
+impl From<Transaction> for CompatTransaction {
+    fn from(value: Transaction) -> Self {
+        match value {
+            Transaction::Legacy(x) => Self::Legacy(x.into()),
+            Transaction::Eip2930(x) => Self::Eip2930(x.into()),
+            Transaction::Eip1559(x) => Self::Eip1559(x.into()),
+            Transaction::Eip4844(x) => Self::Eip4844(x.into()),
+            Transaction::Eip7702(x) => Self::Eip7702(x.into()),
+        }
+    }
+}
+
+impl From<CompatTransaction> for Transaction {
+    fn from(value: CompatTransaction) -> Self {
+        match value {
+            CompatTransaction::Legacy(x) => Self::Legacy(x.into()),
+            CompatTransaction::Eip2930(x) => Self::Eip2930(x.into()),
+            CompatTransaction::Eip1559(x) => Self::Eip1559(x.into()),
+            CompatTransaction::Eip4844(x) => Self::Eip4844(x.into()),
+            CompatTransaction::Eip7702(x) => Self::Eip7702(x.into()),
+        }
+    }
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+struct CompatTransactionSigned {
+    pub hash: [u8; 32],
+    pub signature: CompatSignature,
+    pub transaction: CompatTransaction,
+}
+
+impl From<TransactionSigned> for CompatTransactionSigned {
+    fn from(value: TransactionSigned) -> Self {
+        Self {
+            hash: value.hash.into(),
+            signature: value.signature.into(),
+            transaction: value.transaction.into(),
+        }
+    }
+}
+
+impl From<CompatTransactionSigned> for TransactionSigned {
+    fn from(value: CompatTransactionSigned) -> Self {
+        Self {
+            hash: value.hash.into(),
+            signature: value.signature.into(),
+            transaction: value.transaction.into(),
+        }
+    }
+}
+
+fn compat_to_orig(txs: Vec<CompatTransactionSigned>) -> Vec<TransactionSigned> {
+    fn assert_type<A, B>() {
+        debug_assert_eq!(size_of::<A>(), size_of::<B>());
+        debug_assert_eq!(align_of::<A>(), align_of::<B>());
+    }
+    assert_type::<TransactionSigned, CompatTransactionSigned>();
+    assert_type::<TxLegacy, CompatTxLegacy>();
+    assert_type::<TxEip2930, CompatTxEip2930>();
+    assert_type::<TxEip1559, CompatTxEip1559>();
+    assert_type::<TxEip4844, CompatTxEip4844>();
+    assert_type::<TxEip7702, CompatTxEip7702>();
+    assert_type::<Signature, CompatSignature>();
+    assert_type::<SignedAuthorization, CompatSignedAuthorization>();
+    assert_type::<Parity, CompatParity>();
+    assert_type::<AccessList, CompatAccessList>();
+    assert_type::<AccessListItem, CompatAccessListItem>();
+    assert_type::<TxKind, CompatTxKind>();
+
+    // Unfortunately `xs.into_iter().map(Into::into).collect()` is not
+    //  optimized due to rustc limitations.
+
+    unsafe { core::mem::transmute(txs) }
```

Comment:
sic! Unfortunately this is a smelly code due to https://github.com/rust-lang/rust/issues/135962.

Reply:
As I replied:
> For any pair of pairs `(A0, B0)` and `(A1, B1)`, it is permitted for the physical ordering to be `(A0, B0)` and `(B1, A1)`.

Therefore this is unsound.